clients/src/main/java/org/apache/kafka/clients/ClientTelemetryState.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.errors.IllegalClientTelemetryStateException;
+import org.apache.kafka.common.utils.Utils;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * State that helps determine where client exists in the telemetry state i.e. subscribe->wait->push loop.
+ */
+public enum ClientTelemetryState {
```

Comment:
Seems to be an internal class? Should it go into package `org.apache.kafka.clients.common.internal`?

It seems, client package structure is not very strict about it and lot of presumably internal things are in `org.apache.kafka.clients` (also wondering if every in this package should actually be in `common` anyway?), so it might be ok? Or should we just try to make it stricter going forward? (In KS, we are _very_ strict about it.)

Reply:
I think @kirktrue or @philipnee can answer this better as I am also new to java clients. There would be other classes for telemetry be coming so we should define what goes where.

Reply:
Most of the consumer refactor code goes to the internal package because 1. they are still in the prototype stage 2. they don't meant to be used by the public. I honestly don't know what is `clients.common.interal` for, but it seems to be a sensible place to put it.

Reply:
Thanks @philipnee and @mjsax , I have created another packages in `commons` called `telemetry` where I am planning to push most of the code specifically for KIP-714. I think the `commons.internal` is not holding much of internal classes but mostly in the respective packages under `common`.

Reply:
I do believe it is the convention to put the internal bits inside a nested package, so you'd end up with two packages:

* `o.a.k.common.telemetry`
* `o.a.k.common.telemetry.internal`

Reply:
Thanks @kirktrue, I ll create internal also.

Reply:
Sounds good. Thx.